src/node_api.cc

```diff
@@ -706,6 +706,40 @@ bool FindWrapper(v8::Local<v8::Object> obj,
   return true;
 }
 
+static void DeleteEnv(const v8::WeakCallbackInfo<napi_env__>& data) {
+  delete data.GetParameter();
+}
+
+napi_env get_env(v8::Local<v8::Context> context) {
+  napi_env result;
+
+  auto isolate = context->GetIsolate();
+  auto global = context->Global();
+
+  auto maybe_key = v8::String::NewFromOneByte(isolate,
+      reinterpret_cast<const uint8_t *>("N-API Environment"),
+      v8::NewStringType::kInternalized);
+  CHECK(!maybe_key.IsEmpty());
+  auto key = v8::Private::ForApi(isolate, maybe_key.ToLocalChecked());
+
+  auto maybe_value = global->GetPrivate(context, key);
+  CHECK(!maybe_value.IsEmpty());
+  auto value = maybe_value.ToLocalChecked();
+
+  if (value->IsExternal()) {
+    result = static_cast<napi_env>(value.As<v8::External>()->Value());
+  } else {
+    result = new napi_env__(isolate);
+    auto external = v8::External::New(isolate, result);
+    auto maybe_set = global->SetPrivate(context, key, external);
+    CHECK(maybe_set.FromMaybe(false));
+    v8::Persistent<v8::External> persistent(isolate, external);
```

Comment:
This works because `~Persistent()` currently doesn't call .Reset() but v8.h indicates that may change in the future.  Not sure what to do about that but I figured you should know.